src/op.rs

```diff
@@ -0,0 +1,1288 @@
+    Literal {
+        /// The value to push.
+        value: u64,
+    },
+    /// Indicate that this piece's location is in the given register.
+    Register {
+        /// The register number.
+        register: u64,
+    },
+    /// Find the value of the given register, add the offset, and then
+    /// push the resulting sum on the stack.
+    RegisterOffset {
+        /// The register number.
+        register: u64,
+        /// The offset to add.
+        offset: i64,
+    },
+    /// Compute the frame base (using `DW_AT_frame_base`), add the
+    /// given offset, and then push the resulting sum on the stack.
+    FrameOffset {
+        /// The offset to add.
+        offset: i64,
+    },
+    /// No operation.
+    Nop,
+    /// Push the object address on the stack.
+    PushObjectAddress,
+    /// Evaluate a DWARF expression as a subroutine.  The expression
+    /// comes from the `DW_AT_location` attribute of the indicated
+    /// DIE.
+    Call {
+        /// The DIE to use.
+        offset: DIEReference,
+    },
+    /// Compute the address of a thread-local variable and push it on
+    /// the stack.
+    TLS,
+    /// Compute the call frame CFA and push it on the stack.
+    CallFrameCFA,
+    /// Terminate a piece.
+    Piece {
+        /// The size of this piece in bits.
+        size_in_bits: u64,
+        /// The bit offset of this piece.  If `None`, then this piece
+        /// was specified using `DW_OP_piece` and should start at the
+        /// next byte boundary.
+        bit_offset: Option<u64>,
+    },
+    /// Represents `DW_OP_implicit_value`.
+    ImplicitValue {
+        /// The implicit value to use.
+        data: &'input [u8],
+    },
+    /// Represents `DW_OP_stack_value`.
+    StackValue,
+}
+
+/// A single location of a piece of the result of a DWARF expression.
+pub enum Location<'input> {
+    /// The piece is empty.  Ordinarily this means the piece has been
+    /// optimized away.
+    Empty,
+    /// The piece is found in a register.
+    Register {
+        /// The register number.
+        register: u64,
+    },
+    /// The piece is a scalar value.
+    Scalar {
+        /// The value.
+        value: u64,
+    },
+    /// The piece is represented by some constant bytes.
+    Bytes {
+        /// The value.
+        value: &'input [u8],
+    },
+}
+
+/// The description of a single piece of the result of a DWARF
+/// expression.
+pub struct Piece<'input> {
+    /// If given, the size of the piece in bits.  If `None`, then the
+    /// piece takes its size from the enclosed location.
+    pub size_in_bits: Option<u64>,
+    /// If given, the bit offset of the piece.  If `None`, then the
+    /// piece starts at the next byte boundary.
+    pub bit_offset: Option<u64>,
+    /// Where this piece is to be found.
+    pub location: Location<'input>,
+}
+
+// A helper function to handle branch offsets.
+fn compute_pc(pc: &[u8], bytecode: &[u8], offset: i16) -> ParseResult<usize> {
```

Comment:
It seems slightly strange that the pc comes in as a slice and the new pc is returned as an offset. Would ergonomics be hurt if it took and returned both as a slice or both as offsets? Is there some other consideration here that trumps consistency?